HW4.py

```diff
@@ -0,0 +1,316 @@
+
+    # Groupby df_bin table to create a Difference with mean table
+
+    df_bin_groupby = df_bin.groupby(("LowerBin_UpperBin"), as_index=False).agg(
+        bin_mean=pd.NamedAgg(column="predictor", aggfunc="mean"),
+        bin_count=pd.NamedAgg(column="predictor", aggfunc="count"),
+    )
+
+    bin_center_list = []
+    for bin_center in df_bin_groupby["LowerBin_UpperBin"]:
+        bin_center_list.append(bin_center.mid)
+
+    df_bin_groupby["BinCenter"] = pd.Series(bin_center_list)
+
+    PopulationMean = (np.sum(X) + np.sum(y)) / (count_row * count_column)
+    df_bin["PopulationMean"] = PopulationMean
+    df_bin_groupby["PopulationMean"] = PopulationMean
+
+    MeanSquaredDiff = (
+        df_bin_groupby["bin_mean"] - df_bin_groupby["PopulationMean"]
+    ) ** 2
+    df_bin_groupby["MeanSquaredDiff"] = MeanSquaredDiff
+
+    print(df_bin_groupby)
+
+    # Square the difference, sum them up and divide by number of bins
+    print(
+        f"THE unWeighted NUMBER IS : {df_bin_groupby['MeanSquaredDiff'].sum() / n_of_bin}"  # noqa
+    )
+
+    fig = px.bar(df_bin_groupby, x="BinCenter", y="bin_count")
+    fig.show()
+
+    # Difference with mean table (weighted)
+    print("***Difference with mean table (weighted)***")
+
+    df_bin_groupby_weighted = df_bin_groupby.copy()
+
+    population_proportion = []
+    for count in df_bin_groupby["bin_count"]:
+        population_proportion.append(count / len(predictor_list))
+
+    df_bin_groupby_weighted["PopulationProportion"] = pd.Series(
+        population_proportion
+    )  # noqa
+    df_bin_groupby_weighted["MeanSquaredDiffWeighted"] = (
+        df_bin_groupby_weighted["MeanSquaredDiff"]
+        * df_bin_groupby_weighted["PopulationProportion"]
+    )
+
+    # Square the difference, sum them up and divide by number of bins
+    print(
+        f"THE Weighted NUMBER IS :\
+        {df_bin_groupby_weighted['MeanSquaredDiffWeighted'].sum() / n_of_bin}"
+    )
+
+    print(df_bin_groupby_weighted)
+
+    fig = px.bar(df_bin_groupby_weighted, x="BinCenter", y="bin_count")
+    fig.show()
+
+    # Random Forest Variable importance ranking
+    print("***Random Forest Variable importance ranking***")
+
+    X_train, X_test, y_train, y_test = train_test_split(X, y)
+    model = RandomForestRegressor()
+    model.fit(X, y)
+
+    # get importance
+    importance = model.feature_importances_
+
+    feature = []
+    score = []
+    for i, v in enumerate(importance):
+        feature.append(data.feature_names[i])
+        score.append(round(v, 5))
+
+    df_ranking_vartype = pd.DataFrame(
+        {"Feature": pd.Series(feature), "Score": pd.Series(score)}
+    )
+    df_ranking_vartype["Variable_type"] = predictor_type
+    df_ranking_vartype_sort = df_ranking_vartype.sort_values(by=["Score"])
+
+    # path that will save the ranking excel file
+    path = "D:\PycharmProjects\BDA696-MuTing\Feature_Importance_and_type.xlsx"  # noqa
+    df_ranking_vartype_sort.to_excel(path, index=False)

```

Comment:
My macbook didn't like your D:\ filepath so I had to overwrite this with my code when I ran it to get it to output something. But once I did that everything worked fine.

This is the start for the file. But you should save the other plots and tables to .html files and then include links to them in your final output file.

Reply:
I had to modify this as well,... but didn't count it against you.